File: src/main/kotlin/taskly/system/notification/NotificationAspect.java
package taskly.system.notification;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.*;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import taskly.system.user.User;

import java.text.MessageFormat;

@Aspect
@Component
public class NotificationAspect {

    @Autowired
    private ExpoNotificationService expoNotificationService;

    @Autowired
    private NotificationService notificationService;

    @AfterReturning(
            pointcut = "execution(* *(..)) && @annotation(SendNotification)",
            returning = "result")
    public ExpoNotificationService.DataExpoNotification sendNotification (
            final JoinPoint joinPoint, final Object result) {
        MethodSignature signature = (MethodSignature) joinPoint.getSignature();
        String message = signature.getMethod().getAnnotation(SendNotification.class).value();
        return (result != null) ? sendNotificationToUser(message,
                new NotificationMessageFromObject(result).toString()) : null;
    }

    private ExpoNotificationService.DataExpoNotification sendNotificationToUser(
            final String messageTemplate, final String result) {
        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        String finalMessage = MessageFormat.format(messageTemplate, result);
        Notification notification = new Notification("Notification", finalMessage, user);
        notificationService.save(notification);
        return !user.getExpoNotificationToken().equals("") ?
                expoNotificationService.send(notification) : null;
    }

}

File: src/main/kotlin/taskly/system/notification/NotificationAspect.kt
package taskly.system.notification

import com.sun.xml.internal.ws.spi.db.BindingContextFactory.LOGGER
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.aspectj.lang.annotation.Pointcut
import org.springframework.stereotype.Component

@Aspect
@Component
open class NotificationAspect {

    @Pointcut("within(@org.springframework.stereotype.Repository *)")
    fun repositoryClassMethods() {}

    @Before("repositoryClassMethods()")
    fun beforeSampleCreation() {
        println("aspect code")
    }
}